cli/internal/cacheitem/cacheitem.go

```diff
@@ -0,0 +1,76 @@
+// Package cacheitem is an abstraction over the creation and restoration of a cache
+package cacheitem
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/sha512"
+	"errors"
+	"hash"
+	"io"
+	"log"
+	"os"
+	"reflect"
+	"sync"
+
+	"github.com/vercel/turborepo/cli/internal/turbopath"
+)
+
+var (
+	errMissingSymlinkTarget = errors.New("symlink restoration is delayed")
+	errCycleDetected        = errors.New("links in the cache are cyclic")
+	errTraversal            = errors.New("tar attempts to write outside of directory")
+	errNameMalformed        = errors.New("file name is malformed")
+	errNameWindowsUnsafe    = errors.New("file name is not Windows-safe")
+	errUnsupportedFileType  = errors.New("attempted to restore unsupported file type")
+)
+
+// CacheItem is a `tar` utility with a little bit extra.
+type CacheItem struct {
+	// Path is the location on disk for the CacheItem.
+	Path turbopath.AbsoluteSystemPath
+	// Anchor is the position on disk at which the CacheItem will be restored.
+	Anchor turbopath.AbsoluteSystemPath
+
+	// For creation.
+	once   sync.Once
```

Comment:
Are we expecting this to be used in a multi-threaded environment? Is the rest of this threadsafe?

Reply:
`AddFile` should be safe; so I run the delayed `init` synchronously once. I could have moved it into `Create` but we currently do those thing very far apart from each other and I elected delayed initialization.

Reply:
This is going to disappear; the final approach makes more sense to trigger it once.